sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt

```diff
@@ -1,65 +1,436 @@
+                executePermissionsBlock()
             }
-            field = value
         }
+    }
+
+    private fun loadStoreProductsIfPossible(launchResult: QLaunchResult) {
+        loadStoreProductsIfPossible(launchResult, null, null)
+    }
 
-    lateinit var converter: PurchaseConverter<Pair<SkuDetails, Purchase>>
+    private fun loadStoreProductsIfPossible(launchResult: QLaunchResult,
+                                    onLoadCompleted: ((products: List<SkuDetails>) -> Unit)?,
+                                    onLoadFailed: ((error: BillingError) -> Unit)?) {
+        val productStoreIds = launchResult.products.values.mapNotNull {
+            it.storeID
+        }.toSet()
+        if (!isProductsLoaded && !productStoreIds.isNullOrEmpty()) {
+            billingService.loadProducts(productStoreIds,
+                onLoadCompleted = { details ->
+                    isProductsLoaded = true
+                    val formattedDetails: Map<String, SkuDetails> = configureSkuDetails(details)
+                    skuDetails = formattedDetails.toMutableMap()
+
+                    executeProductsBlocks()
+
+                    onLoadCompleted?.let { it(details) }
+                },
+                onLoadFailed = { error ->
+                    onLoadFailed?.let { it(error) }
+                })
+        }
+    }
+
+    private fun executeProductsBlocks() {
+        if (productsCallbacks.isEmpty()) {
+            return
+        }
+
+        launchResult?.products?.let { products ->
+            products.values.forEach { product ->
+                product.skuDetail = skuDetails[product.storeID]
+            }
+
+            productsCallbacks.forEach {
+                it.onSuccess(products)
+            }
+
+            productsCallbacks.clear()
+        }
+    }
+
+    private fun executePermissionsBlock() {
+        if (permissionsCallbacks.isEmpty()) {
+            return
+        }
+
+        launchResult?.let {
+            val permissions = launchResult?.permissions ?: mapOf()
+            permissionsCallbacks.forEach {
+                it.onSuccess(permissions)
+            }
+        } ?: run {
+            permissionsCallbacks.forEach {
+                val error = launchError ?: QonversionError(QonversionErrorCode.LaunchError)
+                it.onError(error)
+            }
+        }
+
+        permissionsCallbacks.clear()
+    }
+
+    private fun productForID(id: String?): QProduct? {
+        return launchResult?.products?.get(id)
+    }
+
+    private fun consumePurchases(purchases: List<Purchase>) {
+        if (isObserveMode) {
+            return
+        }
+
+        purchases.forEach {purchase ->
+            val skuDetail = skuDetails[purchase.sku]
+            skuDetail?.let { sku ->
+                if (purchase.purchaseState == Purchase.PurchaseState.PENDING) {
+                    return
+                }
+                consume(sku.type, purchase.purchaseToken, purchase.isAcknowledged)
+            }
```

Comment:
You can do like this:
 if (purchase.purchaseState != Purchase.PurchaseState.PENDING) {
                consume(sku.type, purchase.purchaseToken, purchase.isAcknowledged)
            }